### PR TITLE
Added Travis CI Configuration

### LINKED ISSUE #80: Pass pylint with no messages
Right now there's a variety of warnings which ought to get fixed.

### LINKED ISSUE #81: Travis-CI integration
At the very least, we ought to have something to run `pyflakes`. Eventually, we should run `pylint` and unit tests as well, once those are ready to go.

### PR DESCRIPTION
**Pull request by brhoades**
WIP: Not sure if this is done, I need Travis to pick up this branch first.
Cool, pyflakes works. I'll add flake8 and do linting for an hour.

Fixes #81 and fixes #80

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+language: python
+python:
+  - "3.6"
+install:
+  - pip install -r requirements.txt
+script:
+  - pyflakes assigner/**/*.py assigner/*.py
```
Comment (LinuxMercedes) [inline .travis.yml]: This should just be `pyflakes assigner` -- `pyflakes` (and `pylint`) take the name of a python module to check.

Comment (brhoades) [reply .travis.yml]: Thanks, this isn't like flake8 at all.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+language: python
+python:
+  - "3.6"
+install:
+  - pip install -r requirements.txt
+script:
+  - pyflakes assigner
+  - pylint assigner
```
Comment (LinuxMercedes) [inline .travis.yml]: Let's leave `pylint` commented out until I fix all the remaining warnings! I'll start working on that but it'll be a bit (got some 5201 stuff to do first).

Comment (brhoades) [reply .travis.yml]: Already did it, sorry. It's soothing as hell.

**Comment by brhoades** [comment]
@LinuxMercedes I'd appreciate it if you could run this branch through a few tests on gitlabs. I can only confirm so much.

File: assigner/baserepo.py [hunk 3]
```diff
@@ -67,40 +66,56 @@ def from_url(cls, url, token):
...
         url = urljoin(url_base, "/api/v4" + path)
         r = requests.get(url, params=params)
         r.raise_for_status()
         return r.json()
 
     @classmethod
-    def _cls_gl_post(cls, url_base, path, token, payload={}, params={}):
+    def _cls_gl_post(cls, url_base, path, token, payload=None, params=None):
         """Make a Gitlab POST request"""
+        if not params:
```
Comment (brhoades) [inline assigner/baserepo.py]: I'd love a better way to do this. Per pylint, default parameters that are objects are shared across all invocations.

File: assigner/baserepo.py [hunk 4]
```diff
@@ -422,7 +456,7 @@ def new(cls, base_repo, semester, section, username, token):
         return cls.from_url(result["http_url_to_repo"], token)
 
     @classmethod
-    def name(cls, semester, section, assignment, user):
+    def build_name(cls, semester, section, assignment, user):
```
Comment (brhoades) [inline assigner/baserepo.py]: This method had some weird conflict with Repo.name set in __init__. I highly doubt this ever caused a bug.

File: requirements.txt [hunk 5]
```diff
@@ -14,3 +13,7 @@ smmap==0.9.0
 wheel==0.24.0
 PTable==0.9.2
 progressbar2==3.10.1
+pylint==1.8.1
+wrapt==1.10.11
+python-utils==2.2.0
+tqdm==4.19.5
```
Comment (brhoades) [inline requirements.txt]: This was missing :|

File: assigner/baserepo.py [hunk 6]
```diff
@@ -436,18 +470,3 @@ def name(cls, semester, section, assignment, user):
     def push(self, base_repo, branch="master"):
         """Push base_repo code to this repo"""
         base_repo.push_to(self, branch)
-
-
-if __name__ == "__main__":
```
Comment (brhoades) [inline assigner/baserepo.py]: I assumed this wasn't being used and was for testing / from the old one file assigner.
@LinuxMercedes @michaelwisely  confirm?

File: assigner/baserepo.py (hunk 4 above)
Comment (brhoades) [reply assigner/baserepo.py]: It wouldn't work anyway... functions were missing args.

File: assigner/baserepo.py (hunk 6 above)
Comment (LinuxMercedes) [reply assigner/baserepo.py]: Yep, that all can go!

File: .travis.yml (hunk 2 above)
Comment (LinuxMercedes) [reply .travis.yml]: You're a madman!

File: assigner/baserepo.py (hunk 3 above)
Comment (LinuxMercedes) [reply assigner/baserepo.py]: That is true, although in this case it's not a problem and perhaps even cuts down on the amount of garbage allocated (although frankly I'm really not worried about that for Assigner :smile:).

The reason why it's not a problem is that this function doesn't carry state in the default `params` or `payload` objects -- the default `payload` object just stays empty, and the default `params` object just has the `private_token` key which is updated every time around. I suppose we could worry about `requests.post` modifying them but it doesn't seem to right now.

In short: I'm ok with telling `pylint` to ignore that error in these cases.

File: assigner/commands/get.py [hunk 7]
```diff
@@ -70,17 +70,23 @@ def get(conf, args):
...
+                            )
+                        ])
                         row = sec = sid = name = "" # don't print user info more than once
 
                     elif result.old_commit is not None:
-                        output.add_row([row, sec, sid, name, "{}: {} -> {}".format(
-                            result.ref.name, str(result.old_commit)[:8], str(result.ref.commit)[:8]
-                        )])
+                        output.add_row([
+                            row, sec, sid, name, "{}: {} -> {}".format(
```
Comment (LinuxMercedes) [inline assigner/commands/get.py]: Every time I see this line, the `{}: {} -> {}` bit fills me with hope that I've somehow stumbled into a haskell codebase

File: requirements.txt (hunk 5 above)
Comment (LinuxMercedes) [reply requirements.txt]: Ah dang, can you add a step in the release steps in `CONTRIBUTING.md` for updating `requirements.txt`?

**Comment by LinuxMercedes** [comment]
Holy crap this is nice! I'll run it through its paces tomorrow morning.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,7 @@
 ## Devel
 
+- Removed lint, removed old baserepo standalone code, and added travis config.
```
Comment (LinuxMercedes) [inline CHANGELOG.md]: Want to split this into two things? (lint/old code & travis)

I'm no stickler for "one change per PR"

File: assigner/baserepo.py (hunk 3 above)
Comment (brhoades) [reply assigner/baserepo.py]: Awesome, I will go back and update this.

File: assigner/baserepo.py [hunk 9]
```diff
@@ -332,27 +331,30 @@ def archive(self):
...
+        return self._gl_put("/projects/{}/repository/branches/{}/unprotect"
+                            .format(self.id, branch))
 
     def _gl_get(self, path, params={}):
         return self.__class__._cls_gl_get(
             self.url_base, path, self.token, params
         )
 
-    def _gl_post(self, path, payload={}, params={}):
+    def _gl_post(self, path, payload={}, _=None):
```
Comment (LinuxMercedes) [inline assigner/baserepo.py]: We should actually probably be passing `params` through to `_cls_gl_post` here...

File: pylintrc [hunk 10]
```diff
@@ -143,6 +143,7 @@ disable=print-statement,
         too-many-statements,
         attribute-defined-outside-init,
         redefined-builtin,
+        dangerous-default-value,
```
Comment (LinuxMercedes) [inline pylintrc]: I'd like to keep this warning on by default; in general it's correct. In `baserepo.py` we can do something like `#pylint: disable=dangerous-default-value` to silence just those warnings.

**Comment by LinuxMercedes** [comment]
I've poked this some and it seems to work just fine! I'll probably run it through its paces on 5201 homework 1 before I'm comfortable merging it, though.

@brhoades does this fix any bugs or is it "just" syntax changes? If it's the latter, I will probably just roll this into the next release, whenever that happens.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -21,6 +21,7 @@ Before merging:
 - Describe the change in `CHANGELOG.md`
 - Make sure `pyflakes assigner` passes with no errors/warnings
 - Check `pylint assigner` and make sure it's not too egregious. (Eventually we'll get the code to a point where there are no messages from `pylint`...)
```
Comment (LinuxMercedes) [inline CONTRIBUTING.md]: We should change this to require `pylint` to pass now that you've done all the work!!!

Comment (brhoades) [reply CONTRIBUTING.md]: There was some bug pylint claimed to find where StudentRepo.name() was being overriden by Repo.name (the attribute)... but I'm pretty sure this was a nonissue.

File: assigner/commands/get.py (hunk 7 above)
Comment (brhoades) [reply assigner/commands/get.py]: Yeah, it looks like a js lambda for me :D

File: assigner/baserepo.py (hunk 9 above)
Comment (brhoades) [reply assigner/baserepo.py]: Yep, done. This is when happens when you fix lint without reading code.

File: pylintrc (hunk 10 above)
Comment (brhoades) [reply pylintrc]: Done. I agree.

**Comment by brhoades** [comment]
There was some bug that pylint claimed to find where BaseRepo.name = "" was overriding StudentRepo.name(), but since the latter is a classmethod this was a nonissue.

File: CONTRIBUTING.md (hunk 11 above)
Comment (brhoades) [reply CONTRIBUTING.md]: Done.

**Comment by LinuxMercedes** [comment]
@brhoades Can you clean the commit history up some and then let's merge this!

**Comment by brhoades** [comment]
I'll just squash on merge.
